### PR TITLE
Prevent stack overflow when Maven project is imported with Gradle wizard

### PR DESCRIPTION
**Pull request by donat**

### CODE REVIEW

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/internal/DefaultExternalLaunchConfigurationManager.java [hunk 1]
```diff
@@ -170,14 +170,26 @@ private void updateLaunchConfiguration(ILaunchConfigurationWorkingCopy configura
      */
     private class LaunchConfigurationListener implements ILaunchConfigurationListener, EventListener {
 
+        // If another ILaunchConfigurationListener changes the target launch configuration then
+        // a new change event is generated. That event is handled synchronously invoking this
+        // listener again and causing a stack overflow error.
+        // see https://github.com/eclipse/buildship/issues/617
+        private boolean configChangeCalled;
```
Comment (oehme) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/internal/DefaultExternalLaunchConfigurationManager.java]: Is this called from different threads? If yes it needs to be thread safe.

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/internal/DefaultExternalLaunchConfigurationManager.java [hunk 2]
```diff
@@ -170,14 +170,26 @@ private void updateLaunchConfiguration(ILaunchConfigurationWorkingCopy configura
...
 
         @Override
         public void launchConfigurationChanged(ILaunchConfiguration configuration) {
-            updateClasspathProvider(configuration);
+            if (this.configChangeCalled) {
+                return;
+            } else {
+                this.configChangeCalled = true;
+                updateClasspathProvider(configuration);
+                this.configChangeCalled = false;
```
Comment (oehme) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/internal/DefaultExternalLaunchConfigurationManager.java]: Needs to be in a finally block. Otherwise any error will put us into noop mode forever.

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/internal/DefaultExternalLaunchConfigurationManager.java (hunk 1 above)
Comment (donat) [reply org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/internal/DefaultExternalLaunchConfigurationManager.java]: My initial assumption was that it doesn't need to be thread safe because the problem happened on a single thread. But, I've just discovered that that listener can be called from a different thread so I introduced a `ReentrantLock` to guard the execution.

TLDR; fixed.

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/internal/DefaultExternalLaunchConfigurationManager.java (hunk 2 above)
Comment (donat) [reply org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/internal/DefaultExternalLaunchConfigurationManager.java]: Fixed.

**Comment by donat** [comment]
@oehme I've addressed your review items. Would you mind having a second look?

**Comment by donat** [comment]
I'm going to merge this PR since it's blocking the 2.2 release. The impact is minimal and all pending review items are already fixed.

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/internal/DefaultExternalLaunchConfigurationManager.java [hunk 3]
```diff
@@ -170,14 +170,29 @@ private void updateLaunchConfiguration(ILaunchConfigurationWorkingCopy configura
...
+
         @Override
         public void launchConfigurationAdded(ILaunchConfiguration configuration) {
             updateClasspathProvider(configuration);
         }
 
         @Override
-        public void launchConfigurationChanged(ILaunchConfiguration configuration) {
-            updateClasspathProvider(configuration);
+        public synchronized void launchConfigurationChanged(ILaunchConfiguration configuration) {
```
Comment (oehme) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/internal/DefaultExternalLaunchConfigurationManager.java]: We don't want to prevent multiple threads from updating at the same time, we just want to prevent them from overriding each other's state - A threadlocal would be the better choice here I think.

Comment (donat) [reply org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/internal/DefaultExternalLaunchConfigurationManager.java]: I agree. I've fixed it on `master`: https://github.com/eclipse/buildship/commit/1668613eb93e0c9590e38bdd9fdb8249c9431d0f
